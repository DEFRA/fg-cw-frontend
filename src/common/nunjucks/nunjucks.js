import Vision from "@hapi/vision";
import Nunjucks from "nunjucks";

import { glob } from "node:fs/promises";
import { config } from "../config.js";
import { filters } from "../format/format-filters.js";
import { context } from "./context/context.js";
import { formatDate } from "./filters/format-date.js";

const njkPaths = [
  "node_modules/govuk-frontend/dist/",
  ...(await Array.fromAsync(
    glob(["**/views", "**/*(layouts|components|pages|partials)/"], {
      exclude: ["node_modules", "*.*"],
    }),
  )),
];

const viewPaths = await Array.fromAsync(
  glob(["**/views"], {
    exclude: ["node_modules", "*.*"],
  }),
);

export const environment = Nunjucks.configure(njkPaths, {
  autoescape: true,
  throwOnUndefined: false,
  trimBlocks: true,
  lstripBlocks: true,
  watch: config.get("nunjucks.watch"),
  noCache: config.get("nunjucks.noCache"),
  filter: {
    formatDate,
  },
});

environment.addFilter("formatDate", formatDate);
environment.addFilter("fixed", filters.fixed);

export const nunjucks = {
  plugin: Vision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = Nunjucks.compile(src, options.environment);
          return (ctx) => template.render(ctx);
        },
      },
    },
    compileOptions: {
      environment,
    },
    relativeTo: config.get("root"),
    path: viewPaths,
    isCached: config.get("isProduction"),
    context,
  },
};
