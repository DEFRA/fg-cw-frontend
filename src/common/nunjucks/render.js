import camelCase from "lodash/camelCase.js";
import { environment } from "./nunjucks.js";

/**
 * @param {string} name
 * @param {object} params
 * @param {string} [block]
 */
export const render = (name, params, block) => {
  const macro = {
    path: `${name}/macro.njk`,
    name: camelCase(name.split("-")),
    params: JSON.stringify(params, null, 2),
  };

  const callComponent = block
    ? `{%- call ${macro.name}(${macro.params}) -%}${block}{%- endcall -%}`
    : `{{- ${macro.name}(${macro.params}) -}}`;

  return environment.renderString(
    `{%- from "${macro.path}" import ${macro.name} -%}
    ${callComponent}
    `,
    {},
  );
};
