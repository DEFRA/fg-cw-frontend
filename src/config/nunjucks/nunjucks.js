import { fileURLToPath } from 'node:url'
import path from 'path'
import nunjucks from 'nunjucks'
import hapiVision from '@hapi/vision'

import { config } from '../config.js'
import { context } from './context/context.js'
import * as filters from './filters/filters.js'
import * as globals from './globals.js'

const dirname = path.dirname(fileURLToPath(import.meta.url))
const nunjucksEnvironment = nunjucks.configure(
  [
    'node_modules/govuk-frontend/dist/',
    path.resolve(dirname, '../../server/common/templates'),
    path.resolve(dirname, '../../server/common/components'),
    path.resolve(dirname, '../../case-management/views'),
    path.resolve(dirname, '../../case-management/views/layouts'),
    path.resolve(dirname, '../../case-management/views/components'),
    path.resolve(dirname, '../../case-management/views/pages'),
    path.resolve(dirname, '../../case-management/views/partials'),
    path.resolve(dirname, '../../common/presentation')
  ],
  {
    autoescape: true,
    throwOnUndefined: false,
    trimBlocks: true,
    lstripBlocks: true,
    watch: config.get('nunjucks.watch'),
    noCache: config.get('nunjucks.noCache')
  }
)

export const nunjucksConfig = {
  plugin: hapiVision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = nunjucks.compile(src, options.environment)
          return (ctx) => template.render(ctx)
        }
      }
    },
    compileOptions: {
      environment: nunjucksEnvironment
    },
    relativeTo: path.resolve(dirname, '../..'),
    path: ['server', 'case-management/views', 'common/presentation'],
    isCached: config.get('isProduction'),
    context
  }
}

Object.entries(globals).forEach(([name, global]) => {
  nunjucksEnvironment.addGlobal(name, global)
})

Object.entries(filters).forEach(([name, filter]) => {
  nunjucksEnvironment.addFilter(name, filter)
})
