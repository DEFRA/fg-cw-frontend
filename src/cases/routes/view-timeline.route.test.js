import hapi from "@hapi/hapi";
import { load } from "cheerio";
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import { createMockLinks } from "../../../test/data/case-test-data.js";
import { nunjucks } from "../../common/nunjucks/nunjucks.js";
import { findCaseByIdUseCase } from "../use-cases/find-case-by-id.use-case.js";
import { timelineRoute } from "./view-timeline.route.js";

vi.mock("../use-cases/find-case-by-id.use-case.js");

describe("timelineRoute", () => {
  let server;

  beforeAll(async () => {
    server = hapi.server();
    server.route(timelineRoute);
    await server.register([nunjucks]);

    await server.initialize();
  });

  afterAll(async () => {
    await server.stop();
  });

  it("displays a timeline", async () => {
    findCaseByIdUseCase.mockResolvedValue({
      _id: "68495db5afe2d27b09b2ee47",
      caseRef: "banana-123",
      workflowCode: "frps-private-beta",
      status: "NEW",
      dateReceived: "2025-06-11T10:43:01.603Z",
      currentStage: "application-receipt",
      links: createMockLinks("68495db5afe2d27b09b2ee47"),
      payload: {
        clientRef: "banana-123",
        code: "frps-private-beta",
        createdAt: "2025-06-11T10:43:01.417Z",
        submittedAt: "2023-10-01T12:00:00.000Z",
        identifiers: {
          sbi: "SBI001",
          frn: "FIRM0001",
          crn: "CUST0001",
          defraId: "DEFRA0001",
        },
        answers: {
          agreementName: "Test application name 1",
          scheme: "SFI",
          year: 2025,
          hasCheckedLandIsUpToDate: true,
          actionApplications: [
            {
              parcelId: "9238",
              sheetId: "SX0679",
              code: "CSAM1",
              appliedFor: {
                unit: "ha",
                quantity: 20.23,
              },
            },
          ],
        },
      },
      stages: [
        {
          code: "application-receipt",
          title: "Application Receipt",
          taskGroups: [
            {
              id: "application-receipt-tasks",
              title: "Application Receipt Tasks",
              tasks: [
                {
                  id: "simple-review",
                  title: "Simple Review",
                  status: "pending",
                },
              ],
            },
          ],
        },
        {
          code: "contract",
          title: "Contract",
          taskGroups: [],
        },
      ],
      timeline: [
        {
          eventType: "CASE_CREATED",
          createdAt: "2025-06-16T09:01:14.072Z",
          description: "Case received",
          createdBy: { name: "Julian Kimmings" },
          data: {
            caseRef: "APPLICATION-REF-3",
          },
        },
        {
          eventType: "CASE_ASSIGNED",
          description: "Case assigned",
          createdAt: "2025-06-16T09:01:14.072Z",
          createdBy: { name: "Julian Kimmings" },
          data: {
            assignedTo: { name: "martin smith" },
          },
        },
        {
          eventType: "TASK_COMPLETED",
          description: "Task completed",
          createdAt: "2025-06-16T09:01:14.072Z",
          createdBy: { name: "Nicholai Hel" },
          data: {
            taskGroupId: "tg-01",
            taskId: "task-01",
          },
        },
      ],
    });

    const { statusCode, result } = await server.inject({
      method: "GET",
      url: "/cases/68495db5afe2d27b09b2ee47/timeline",
      auth: {
        credentials: { token: "mock-token" },
        strategy: "session",
      },
    });

    expect(statusCode).toEqual(200);

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });
});
