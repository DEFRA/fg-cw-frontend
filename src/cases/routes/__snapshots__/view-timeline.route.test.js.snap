// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`timelineRoute > displays a timeline 1`] = `
"
  
<div class="govuk-!-margin-bottom-6 app-case-banner">
  <div class="govuk-width-container">
    <a href="/cases" class="govuk-back-link">Back to cases</a>

    <h1 class="govuk-heading-m">
      Test application name 1
    </h1>

    <div>
      <div class="app-case-row">
        <div class="app-case-cell">
          <p>
            <strong class="govuk-!-margin-right-1">SBI:</strong>
            SBI001
          </p>
        </div>

        <div class="app-case-cell">
          <p>
            <strong class="govuk-!-margin-right-1">Reference:</strong>
            frps-private-beta
          </p>
        </div>

        <div class="app-case-cell">
          <p>
            <strong class="govuk-!-margin-right-1">Scheme:</strong>
            SFI
            2025
          </p>
        </div>

        <div class="app-case-cell">
          <p>
            <strong class="govuk-!-margin-right-1">Date received:</strong>
            11 Jun 2025
          </p>
        </div>

        <div class="app-case-cell">
          <p>
            <strong class="govuk-!-margin-right-1">Status:</strong>
            NEW
          </p>
        </div>
      </div>
    </div>
  </div>
</div>


  <div class="govuk-body">
    
  <div class="govuk-service-navigation" data-module="govuk-service-navigation">
      <div class="govuk-width-container">

    <div class="govuk-service-navigation__container">
      

      
        <nav aria-label="Menu" class="govuk-service-navigation__wrapper">
          <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="navigation" hidden="">
            Menu
          </button>

          <ul class="govuk-service-navigation__list" id="navigation">

            
              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47">
                                    
Tasks
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/case-details">
                                    
Case Details
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item govuk-service-navigation__item--active">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/timeline" aria-current="true">
                                    
                  <strong class="govuk-service-navigation__active-fallback">Timeline</strong>

                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="#">
                                    
Land Details
                  </a>
              </li>

            </ul>
        </nav>
    </div>

    </div>

  </div>



    <h2 class="govuk-heading-l govuk-!-margin-bottom-1 govuk-!-margin-top-6">
      Timeline
    </h2>

    <div class="govuk-grid-row">
      <div class="govuk-grid-column-full">
        <div class="govuk-body" data-testid="app-page-body">
          <div class="timeline govuk-!-margin-top-3">
              <div class="timeline__item">
                <div class="timeline__header">
                  <h2>    Case received</h2>

                  <p class="timeline__byline">by Julian Kimmings</p>
                </div>
                <p class="timeline__date">
                  <time datetime="2025-06-16T09:01:14.072Z">16 Jun 2025 10:01</time>
                </p>
                <div class="timeline__description">
</div>

              </div>
              <div class="timeline__item">
                <div class="timeline__header">
                  <h2>    Case assigned
    to martinUserId</h2>

                  <p class="timeline__byline">by Julian Kimmings</p>
                </div>
                <p class="timeline__date">
                  <time datetime="2025-06-16T09:01:14.072Z">16 Jun 2025 10:01</time>
                </p>
                <div class="timeline__description">
    <p>
      Assigned to martinUserId by nickUserId
    </p>
</div>

              </div>
              <div class="timeline__item">
                <div class="timeline__header">
                  <h2>    Task completed</h2>

                  <p class="timeline__byline">by Nicholai Hel</p>
                </div>
                <p class="timeline__date">
                  <time datetime="2025-06-16T09:01:14.072Z">16 Jun 2025 10:01</time>
                </p>
                <div class="timeline__description">
    <!-- TODO  task link? -->
    <a class="govuk-link" href="/cases//tasks//">View task</a>
</div>

              </div>
          </div>
        </div>
      </div>
    </div>
  </div>
        "
`;
